File: examples/algorithms-python/src/leetcode/lc0133_clone_graph.py
"""
LeetCode 133. Clone Graph
链接: https://leetcode.com/problems/clone-graph/
难度: Medium

题目描述:
给你无向连通图中一个节点的引用，请你返回该图的深拷贝。

形式化规约:
  输入: 连通无向图 G = (V, E) 的一个节点的引用, |V| ∈ [1, 100]
  输出: G' ≅ G 的一个节点的引用, 且 V' ∩ V = ∅

最优解思路:
  DFS/BFS + 哈希映射（引用不变式）：
  使用字典维护 "原节点 → 新节点" 的映射，防止环导致的无限递归。
  每个原节点仅被复制一次。

复杂度:
  时间: O(V + E)
  空间: O(V)
"""

from typing import Optional


class Node:
    def __init__(self, val: int = 0, neighbors: Optional[list["Node"]] = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        """
        DFS 深拷贝图。时间 O(V+E)，空间 O(V)。
        """
        if node is None:
            return None

        visited: dict[int, Node] = {}

        def dfs(n: Node) -> Node:
            if n.val in visited:
                return visited[n.val]

            # 立即建立映射，打破环的循环依赖
            clone = Node(n.val)
            visited[n.val] = clone

            # 递归复制邻居
            for neighbor in n.neighbors:
                clone.neighbors.append(dfs(neighbor))

            return clone

        return dfs(node)


def test_clone_graph():
    def build_graph(adj_list: list[list[int]]) -> Optional[Node]:
        if not adj_list:
            return None
        nodes = {i: Node(i) for i in range(1, len(adj_list) + 1)}
        for i, neighbors in enumerate(adj_list, 1):
            for n in neighbors:
                nodes[i].neighbors.append(nodes[n])
        return nodes[1]

    def graph_to_adj_list(node: Optional[Node]) -> list[list[int]]:
        if node is None:
            return []
        result_map: dict[int, list[int]] = {}
        stack = [node]
        visited_vals = set()
        while stack:
            n = stack.pop()
            if n.val in visited_vals:
                continue
            visited_vals.add(n.val)
            neighbors = []
            for nb in n.neighbors:
                neighbors.append(nb.val)
                stack.append(nb)
            result_map[n.val] = sorted(neighbors)
        return [result_map[i] for i in range(1, max(result_map.keys()) + 1)]

    # 测试 1: 基本图
    adj1 = [[2, 4], [1, 3], [2, 4], [1, 3]]
    original1 = build_graph(adj1)
    cloned1 = Solution().cloneGraph(original1)
    assert graph_to_adj_list(cloned1) == adj1, "Test 1 failed"

    # 测试 2: 单节点
    adj2 = [[]]
    original2 = build_graph(adj2)
    cloned2 = Solution().cloneGraph(original2)
    assert graph_to_adj_list(cloned2) == adj2, "Test 2 failed"

    # 测试 3: 空图
    assert Solution().cloneGraph(None) is None, "Test 3 failed"

    print("All tests passed for LC 133 - Clone Graph")


if __name__ == "__main__":
    test_clone_graph()

File: examples/algorithms/src/leetcode/lc0133_clone_graph.rs
//! LeetCode 133. Clone Graph
//! 链接: https://leetcode.com/problems/clone-graph/
//! 难度: Medium
//!
//! 算法: DFS/BFS + 哈希映射（引用不变式）
//! 时间复杂度: O(V + E)
//! 空间复杂度: O(V)

use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

#[derive(Debug, Clone)]
pub struct Node {
    pub val: i32,
    pub neighbors: Vec<Rc<RefCell<Node>>>,
}

impl Node {
    pub fn new(val: i32) -> Rc<RefCell<Node>> {
        Rc::new(RefCell::new(Node {
            val,
            neighbors: Vec::new(),
        }))
    }
}

pub struct Solution;

impl Solution {
    /// 返回给定连通无向图的深拷贝。
    ///
    /// 形式化规约:
    /// - 前置条件: G = (V, E) 是连通无向图, |V| ∈ [1, 100]
    /// - 后置条件: 返回 G' ≅ G, 且 V' ∩ V = ∅（所有节点为新创建）
    ///
    /// 核心思路:
    ///   使用哈希表维护 "原节点 → 新节点" 的映射，防止环导致的无限递归。
    ///   DFS 遍历原图，每个原节点仅被复制一次。
    pub fn clone_graph(node: Option<Rc<RefCell<Node>>>) -> Option<Rc<RefCell<Node>>> {
        if node.is_none() {
            return None;
        }

        let mut visited: HashMap<i32, Rc<RefCell<Node>>> = HashMap::new();
        Self::dfs(node.unwrap(), &mut visited)
    }

    fn dfs(
        node: Rc<RefCell<Node>>,
        visited: &mut HashMap<i32, Rc<RefCell<Node>>>,
    ) -> Option<Rc<RefCell<Node>>> {
        let val = node.borrow().val;

        // 引用不变式: 若已复制，直接返回缓存的新节点
        if let Some(clone) = visited.get(&val) {
            return Some(clone.clone());
        }

        // 创建新节点并立即建立映射（打破环的循环依赖）
        let clone = Node::new(val);
        visited.insert(val, clone.clone());

        // 递归复制邻居
        for neighbor in &node.borrow().neighbors {
            if let Some(clone_neighbor) = Self::dfs(neighbor.clone(), visited) {
                clone.borrow_mut().neighbors.push(clone_neighbor);
            }
        }

        Some(clone)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn build_graph(adj_list: Vec<Vec<i32>>) -> Option<Rc<RefCell<Node>>> {
        if adj_list.is_empty() {
            return None;
        }

        let mut nodes: HashMap<i32, Rc<RefCell<Node>>> = HashMap::new();
        for i in 1..=adj_list.len() {
            nodes.insert(i as i32, Node::new(i as i32));
        }

        for (i, neighbors) in adj_list.iter().enumerate() {
            let val = (i + 1) as i32;
            for &n in neighbors {
                nodes
                    .get(&val)
                    .unwrap()
                    .borrow_mut()
                    .neighbors
                    .push(nodes.get(&n).unwrap().clone());
            }
        }

        Some(nodes.get(&1).unwrap().clone())
    }

    fn graph_to_adj_list(node: Option<Rc<RefCell<Node>>>) -> Vec<Vec<i32>> {
        if node.is_none() {
            return vec![];
        }

        let mut map: HashMap<i32, Vec<i32>> = HashMap::new();
        let mut stack = vec![node.unwrap()];
        let mut visited_vals = std::collections::HashSet::new();

        while let Some(n) = stack.pop() {
            let val = n.borrow().val;
            if visited_vals.contains(&val) {
                continue;
            }
            visited_vals.insert(val);

            let mut neighbors = Vec::new();
            for neighbor in &n.borrow().neighbors {
                neighbors.push(neighbor.borrow().val);
                stack.push(neighbor.clone());
            }
            neighbors.sort();
            map.insert(val, neighbors);
        }

        let max_val = *map.keys().max().unwrap();
        let mut result = Vec::new();
        for i in 1..=max_val {
            result.push(map.get(&i).cloned().unwrap_or_default());
        }
        result
    }

    #[test]
    fn test_clone_graph_basic() {
        let adj_list = vec![vec![2, 4], vec![1, 3], vec![2, 4], vec![1, 3]];
        let original = build_graph(adj_list.clone());
        let cloned = Solution::clone_graph(original);
        assert_eq!(graph_to_adj_list(cloned), adj_list);
    }

    #[test]
    fn test_clone_graph_single_node() {
        let adj_list = vec![vec![]];
        let original = build_graph(adj_list.clone());
        let cloned = Solution::clone_graph(original);
        assert_eq!(graph_to_adj_list(cloned), adj_list);
    }

    #[test]
    fn test_clone_graph_empty() {
        assert!(Solution::clone_graph(None).is_none());
    }
}
